browser/net/brave_proxying_url_loader_factory.cc

```diff
@@ -0,0 +1,969 @@
+  if (result != MOJO_RESULT_OK) {
+    network::URLLoaderCompletionStatus status(net::ERR_FAILED);
+    write_data->client->OnComplete(status);
+    return;
+  }
+
+  network::URLLoaderCompletionStatus status(net::OK);
+  status.encoded_data_length = write_data->data.size();
+  status.encoded_body_length = write_data->data.size();
+  status.decoded_body_length = write_data->data.size();
+  write_data->client->OnComplete(status);
+}
+
+}  // namespace
+
+BraveProxyingURLLoaderFactory::InProgressRequest::FollowRedirectParams::
+    FollowRedirectParams() = default;
+BraveProxyingURLLoaderFactory::InProgressRequest::FollowRedirectParams::
+    ~FollowRedirectParams() = default;
+
+BraveProxyingURLLoaderFactory::InProgressRequest::InProgressRequest(
+    BraveProxyingURLLoaderFactory* factory,
+    uint64_t request_id,
+    int32_t network_service_request_id,
+    int32_t routing_id,
+    int render_process_id,
+    int frame_tree_node_id,
+    uint32_t options,
+    const network::ResourceRequest& request,
+    content::ResourceContext* resource_context,
+    const net::MutableNetworkTrafficAnnotationTag& traffic_annotation,
+    network::mojom::URLLoaderRequest loader_request,
+    network::mojom::URLLoaderClientPtr client)
+    : factory_(factory),
+      request_(request),
+      original_initiator_(request.request_initiator),
+      request_id_(request_id),
+      network_service_request_id_(network_service_request_id),
+      render_process_id_(render_process_id),
+      frame_tree_node_id_(frame_tree_node_id),
+      routing_id_(routing_id),
+      options_(options),
+      resource_context_(resource_context),
+      traffic_annotation_(traffic_annotation),
+      proxied_loader_binding_(this, std::move(loader_request)),
+      target_client_(std::move(client)),
+      proxied_client_binding_(this),
+      has_any_extra_headers_listeners_(network_service_request_id_ != 0),
+      header_client_binding_(this),
+      weak_factory_(this) {
+  // If there is a client error, clean up the request.
+  target_client_.set_connection_error_handler(base::BindOnce(
+      &BraveProxyingURLLoaderFactory::InProgressRequest::OnRequestError,
+      weak_factory_.GetWeakPtr(),
+      network::URLLoaderCompletionStatus(net::ERR_ABORTED)));
+}
+
+BraveProxyingURLLoaderFactory::InProgressRequest::~InProgressRequest() {
+  if (on_before_send_headers_callback_) {
+    std::move(on_before_send_headers_callback_)
+        .Run(net::ERR_ABORTED, base::nullopt);
+  }
+  if (on_headers_received_callback_) {
+    std::move(on_headers_received_callback_)
+        .Run(net::ERR_ABORTED, base::nullopt, GURL());
+  }
+  if (ctx_) {
+    factory_->request_handler_->OnURLRequestDestroyed(ctx_);
+  }
+}
+
+void BraveProxyingURLLoaderFactory::InProgressRequest::Restart() {
+  UpdateRequestInfo();
+  RestartInternal();
+}
+
+void BraveProxyingURLLoaderFactory::InProgressRequest::UpdateRequestInfo() {
+  // TODO(iefremov): Update |ctx_| here and get rid of multiple spots where
```

Comment:
I think this needs to be done before merging

Reply:
not sure... I'd say it needs to be done after writing a solid pack of browser tests, that is not going to happen before C77 will be merged